service.go

```diff
@@ -0,0 +1,541 @@
+				tserv.AddOpt.KeyLog(req.LogID.ID))
+			if err != nil {
+				return nil, err
+			}
+		}
+	} else {
+		// Verify node
+		err = rec.Verify(logpk)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = s.threads.Put(ctx, rec, tserv.PutOpt.Thread(req.ThreadID.ID), tserv.PutOpt.Log(req.LogID.ID))
+	if err != nil {
+		return nil, err
+	}
+
+	return reply, nil
+}
+
+func (s *service) Pull(ctx context.Context, req *pb.PullRequest) (*pb.PullReply, error) {
+	recs, err := s.threads.pullLocal(ctx, req.ThreadID.ID, req.LogID.ID, req.Offset.Cid, int(req.Limit))
+	if err != nil {
+		return nil, err
+	}
+
+	pbrecs := &pb.PullReply{
+		Records: make([]*pb.Record, len(recs)),
+	}
+	for i, r := range recs {
+		pbrecs.Records[i], err = cbor.RecordToProto(ctx, s.threads.dagService, r)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return pbrecs, nil
+}
+
+// push a record to log addresses and thread topic.
+func (s *service) push(ctx context.Context, rec thread.Record, id thread.ID, lid peer.ID, settings *tserv.AddSettings) error {
+	var addrs []ma.Multiaddr
+	// Collect known writers
+	for _, l := range s.threads.ThreadInfo(settings.Thread).Logs {
+		if l.String() == lid.String() {
+			continue
+		}
+		addrs = append(addrs, s.threads.Addrs(settings.Thread, l)...)
+	}
+
+	// Add additional addresses
+	addrs = append(addrs, settings.Addrs...)
+
+	// Serialize and sign the record for transport
+	pbrec, err := cbor.RecordToProto(ctx, s.threads.dagService, rec)
+	if err != nil {
+		return err
+	}
+	payload, err := pbrec.Marshal()
+	if err != nil {
+		return err
+	}
+	sk := s.threads.getPrivKey()
+	if sk == nil {
+		return fmt.Errorf("could not find key for host")
+	}
+	sig, err := sk.Sign(payload)
+	if err != nil {
+		return err
+	}
+
+	var keyLog *pb.ProtoPeerID
+	logKey := s.threads.ReadKey(settings.Thread, settings.KeyLog)
+	if logKey != nil {
+		keyLog = &pb.ProtoPeerID{ID: settings.KeyLog}
+	}
+	req := &pb.PushRequest{
+		Header: &pb.PushRequest_Header{
+			From:         &pb.ProtoPeerID{ID: s.threads.host.ID()},
+			Signature:    sig,
+			Key:          &pb.ProtoPubKey{PubKey: sk.GetPublic()},
+			FollowKey:    s.threads.FollowKey(id, lid),
+			ReadKeyLogID: keyLog,
+		},
+		ThreadID: &pb.ProtoThreadID{ID: id},
+		LogID:    &pb.ProtoPeerID{ID: lid},
+		Record:   pbrec,
+	}
+
+	// Push to each address
+	wg := sync.WaitGroup{}
+	for _, addr := range addrs {
+		wg.Add(1)
+		go func(addr ma.Multiaddr) {
+			p, err := addr.ValueForProtocol(ma.P_P2P)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			pid, err := peer.IDB58Decode(p)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Debugf("pushing record to %s...", p)
+
+			cctx, _ := context.WithTimeout(ctx, reqTimeout)
+			conn, err := s.dial(cctx, pid, grpc.WithInsecure())
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			client := pb.NewThreadsClient(conn)
+			reply, err := client.Push(cctx, req)
+			if err != nil {
+				log.Error(err)
+				return
```

Comment:
Warning! (both returns statements): `wg.Done()` isn't `defer`ed, so this will deadlock the parent-goroutine.

Reply:
Oh! good catch :)